apollo-at-link/src/federation_schema/remove.rs

```diff
@@ -0,0 +1,2131 @@
+use crate::federation_schema::{OptionLinksMetadata, ReferencerFederationSchemaMut};
+use crate::location::{
+    DirectiveArgumentDefinitionLocation, DirectiveDefinitionLocation, EnumTypeDefinitionLocation,
+    EnumValueDefinitionLocation, InputObjectFieldDefinitionLocation,
+    InputObjectTypeDefinitionLocation, InterfaceFieldArgumentDefinitionLocation,
+    InterfaceFieldDefinitionLocation, InterfaceTypeDefinitionLocation,
+    ObjectFieldArgumentDefinitionLocation, ObjectFieldDefinitionLocation,
+    ObjectTypeDefinitionLocation, ScalarTypeDefinitionLocation, SchemaDefinitionLocation,
+    SchemaRootDefinitionKind, SchemaRootDefinitionLocation, UnionTypeDefinitionLocation,
+};
+use crate::referencer::{
+    DirectiveReferencers, EnumTypeReferencers, InputObjectTypeReferencers,
+    InterfaceTypeReferencers, ObjectTypeReferencers, Referencers, ScalarTypeReferencers,
+    UnionTypeReferencers,
+};
+use apollo_compiler::schema::{
+    Component, ComponentStr, Directive, EnumType, EnumValueDefinition, ExtendedType,
+    FieldDefinition, InputObjectType, InputValueDefinition, InterfaceType, Name, ObjectType,
+    ScalarType, UnionType,
+};
+use apollo_compiler::{Node, Schema};
+use indexmap::Equivalent;
+use std::hash::Hash;
+use std::ops::Deref;
+
+impl SchemaDefinitionLocation {
+    pub fn remove_directive_name<
+        T: AsMut<OptionLinksMetadata> + AsRef<OptionLinksMetadata>,
+        U: AsMut<Referencers> + AsRef<Referencers>,
+        Q: Hash + Equivalent<Name>,
+    >(
+        &self,
+        schema: &mut ReferencerFederationSchemaMut<T, U>,
+        name: &Q,
+    ) {
```

Comment:
Instead of removing a tuple, I'd either want this to return its own type, or perhaps even the original struct itself without that directive name. It's a bit strange to have it return something vastly different than the original `SchemaDefinitionLocation` itself.

And is the `Location` struct that should be removing a directive? Not federation schema?

Reply:
For a removal, I'd maybe borrow an API that already exists in rust std. For example, when you insert into a hashmap, and a key already exists, it returns the value of the item that was replaced. this could also return the item that was removed. (maybe even also in an option)

Reply:
this goes for all of these removal functions in this file ^